tests: Support marking topotests as "skip by default"

gromit1811 opened the pull request:
Long-running topotests shouldn't be run by default to avoid delaying CI too
much, so allow them to be marked as "skip by default". In that case, they
get skipped unless the command line option "--runall" is specified.

The following decorator is used to mark a topotest accordingly:
@pytest.mark.skip_by_default

Signed-off-by: Martin Buck <mb-tmp-tvguho.pbz@gromit.dyndns.org>

polychaeta reviewed (commented):
Thanks for your contribution to FRR!


<details>
<summary><b>Click for style suggestions</b></summary>

To apply these suggestions:

<p>

```
curl -s https://gist.githubusercontent.com/polychaeta/6e3ad6672232b3e8cf38dfc445389545/raw/47a22fd2879c0bbc0cab1db7803c0f39ed2b4fd6/cr_7613_1606422155.diff | git apply
```

</p>

<p>

```diff
diff --git a/tests/topotests/conftest.py b/tests/topotests/conftest.py
index 7bc8cea48..ed599f90f 100755
--- a/tests/topotests/conftest.py
+++ b/tests/topotests/conftest.py
@@ -25,7 +25,8 @@ def pytest_addoption(parser):
     parser.addoption(
         "--runall",
         action="store_true",
-        help="Run all tests, including those disabled by default")
+        help="Run all tests, including those disabled by default",
+    )
 
 
 def pytest_runtest_call():
@@ -68,7 +69,9 @@ def pytest_configure(config):
     if config.getoption("--topology-only"):
         topology_only = True
 
-    config.addinivalue_line("markers", "skip_by_default: mark test as to be run only on request")
+    config.addinivalue_line(
+        "markers", "skip_by_default: mark test as to be run only on request"
+    )
 
 
 def pytest_runtest_makereport(item, call):

```

</p>
</details>

            
If you are a new contributor to FRR, please see our [contributing guidelines](http://docs.frrouting.org/projects/dev-guide/en/latest/workflow.html#coding-practices-style).

After making changes, you do not need to create a new PR. You should perform an [amend or interactive rebase](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History) followed by a [force push](https://git-scm.com/docs/git-push#Documentation/git-push.txt---force).

gromit1811 commented:
#7168 is a candidate which could make use of this new feature. This is a topotest which typically needs a long time to run, so it shouldn't run by default. I'll update that PR and mark the new topotest accordingly once this one here is merged.

donaldsharp commented:
I disagree that we should not run some topotests all the time based *only* on run length.  These tests while imperfect are finding issues and I would hate to have something get in that could have been stopped earlier.  If I was going to approach this issue I would have some intelligent bit of code in the CI system where if I make an ospf only change we only run ospf tests.  Or if I change zebra or lib, then all tests a rerun.  If I modify bgp then all bgp tests are run.  That would help tremendously.

gromit1811 commented:
> I disagree that we should not run some topotests all the time based _only_ on run length. These tests while imperfect are finding issues and I would hate to have something get in that could have been stopped earlier. If I was going to approach this issue I would have some intelligent bit of code in the CI system where if I make an ospf only change we only run ospf tests. Or if I change zebra or lib, then all tests a rerun. If I modify bgp then all bgp tests are run. That would help tremendously.

This was requested in https://github.com/FRRouting/frr/pull/7168#pullrequestreview-499147869. That particular topotest checks for a very specific regression that is relatively unlikely to show up again and its main purpose was rather to demonstrate the previously existing bug.

Delaying every PR for an hour due to this is not reasonable, IMO. Delaying only OSPF6-related PRs as you suggest makes more sense. But if you ask me, it's still too much, especially if we have more tests like this in the future.

I don't know what's (easily) possible with our CI system, but wouldn't it make sense to run tests like this decoupled from PRs on a regular basis, e.g. once a week? With automatic bisecting, identifying the guilty commit would still be possible, even if it might have already been merged by then. But if that happens as rarely as I expect (at least in case of #7168), that wouldn't be a big issue IMO.

gromit1811 commented:
Finally found some time for this (sometimes, technical problems delaying a Linux Netdev conf presentation can give you some unplanned spare time to work on something ;-), sorry for the loooong delay.

Rebased to current master, updated to match handling of other options which have been implemented in the meantime and added a `--run-skipped` option as suggested by @mwinter-osr. Hope it looks good to you.

gromit1811 commented:
Looks like there is no interest in this as #7168 which was the only user was closed long ago
